### ISSUE TITLE
Can we use BDD Dataset directly to train model ?

### ISSUE THREAD

**Issue by letdivedeep** [cfzd/Ultra-Fast-Lane-Detection#141]
Hi @cfzd,

Can we use BDD Dataset directly to train model? 

As BDD dataset has max 8 lane and CULane has max 4 lane annotations.
          -

**Comment by cfzd** [comment]
@letdivedeep 
Setting the `num_lanes` to 8 should work. But I think it might be a little tricky to handle with the lane order since BDD seems to have unordered lane annotations.

**Comment by letdivedeep** [comment]
@cfzd 

Do we have any solution to this problem of unordered annotations in BDD?

**Comment by cfzd** [comment]
For this approach, it seems it would be hard.

**Comment by letdivedeep** [comment]
@cfzd  thanks for the reply

It means train the model over BDD dataset may not yield good results, is my understanding right?

**Comment by cfzd** [comment]
@letdivedeep 
In fact, if the lane is not ordered, I even don't think this method would converge. The channels of the output layer are related to different lanes according to some kinds of rules or orders defined by datasets. If you don't have a rule or order, the network simply will not converge. If you want to make this method work on BDD, you might need to define some kinds of orders or rules, so that the network could learn how to output these lanes.

**Comment by letdivedeep** [comment]
@cfzd 

thanks.

**Comment by letdivedeep** [comment]
@cfzd 

 if I want to convert the predicted output from the UFS model to CULane annotation format is possible?

**Comment by cfzd** [comment]
@letdivedeep 
Yes, it is possible. For CULane, the segmentation map has four classes, which are left-left lane, left lane, right lane, right-right lane, and the indexes are 1,2,3,4 correspondingly. So you need to do:
- get the predicted results, i.e. the coordinates of lane points, of the four lanes
- draw the lane lines on an empty canvas.
  You can use cv2.line and NumPy to achieve this. cv2.line has a parameter called color, you can use this to make your segmentation map has indexes of 1,2,3,4, e.g. cv2.line(..., color = (1,1,1))
- store the segmentation map
  Here exist a trick, since the segmentation map only contains lane label and is not colored, you can store only one channel of segmentation, it still works.

**Comment by letdivedeep** [comment]
@cfzd 

Thanks for the detailed steps. I will try this out

**Comment by letdivedeep** [comment]
@cfzd ... I have implemented the above suggestion 

I am getting the below error, seam like something going wrong while calculating the loss 

![Screenshot 2020-12-16 at 11 04 19 PM](https://user-images.githubusercontent.com/74127861/102384905-37c21080-3ff3-11eb-9ddb-8e1c0f787aa6.png)

steps followed : 
> Infer images of shape(512,256) over the pre-trained tusimple model
> get the predicted results
> draw the lane lines on NumPy canvas of single dim (where segmentation map has indexes of 1,2,3,4,)
> write the NumPy canvas in dir
> run the training script
for generating segme5ntation map 

Note: the image and seg image is saved of shape 512x256

Any Pointer would be helpful

**Comment by cfzd** [comment]
@letdivedeep 
The `device-side assert triggered` error is usually related to the wrong label exceeds the range of output. For example, you have a 1000-class classification network, but the label is larger than 1000, say 1001, then this error occurs.

**Comment by letdivedeep** [comment]
@cfzd thanks for the reply

I was doing the debugging, what I found was when i set the use_aux= false. Training works without seg loss

seams like an issue with the seg map
